datafusion/datasource/src/url.rs

```diff
@@ -242,33 +242,64 @@ impl ListingTableUrl {
-            true => match ctx.runtime_env().cache_manager.get_list_files_cache() {
-                None => store.list(Some(&self.prefix)),
+
+        async fn list_with_cache<'b>(
+            ctx: &'b dyn Session,
+            store: &'b dyn ObjectStore,
+            prefix: &'b Path,
+        ) -> Result<BoxStream<'b, Result<ObjectMeta>>> {
+            match ctx.runtime_env().cache_manager.get_list_files_cache() {
+                None => Ok(store
+                    .list(Some(prefix))
+                    .map(|res| res.map_err(|e| DataFusionError::ObjectStore(Box::new(e))))
+                    .boxed()),
                 Some(cache) => {
-                    if let Some(res) = cache.get(&self.prefix) {
+                    if let Some(res) = cache.get(prefix) {
                         debug!("Hit list all files cache");
-                        futures::stream::iter(res.as_ref().clone().into_iter().map(Ok))
-                            .boxed()
+                        Ok(futures::stream::iter(
+                            res.as_ref().clone().into_iter().map(Ok),
+                        )
+                        .map(|res| {
+                            res.map_err(|e| DataFusionError::ObjectStore(Box::new(e)))
+                        })
+                        .boxed())
                     } else {
-                        let list_res = store.list(Some(&self.prefix));
-                        let vec = list_res.try_collect::<Vec<ObjectMeta>>().await?;
-                        cache.put(&self.prefix, Arc::new(vec.clone()));
-                        futures::stream::iter(vec.into_iter().map(Ok)).boxed()
+                        let vec = store
+                            .list(Some(prefix))
+                            .map(|res| {
+                                res.map_err(|e| DataFusionError::ObjectStore(Box::new(e)))
+                            })
+                            .try_collect::<Vec<ObjectMeta>>()
+                            .await?;
+                        cache.put(prefix, Arc::new(vec.clone()));
+                        Ok(futures::stream::iter(vec.into_iter().map(Ok))
+                            .map(|res| {
+                                res.map_err(|e| DataFusionError::ObjectStore(Box::new(e)))
+                            })
+                            .boxed())
                     }
                 }
-            },
-            false => futures::stream::once(store.head(&self.prefix)).boxed(),
+            }
+        }
+
+        let list: BoxStream<'a, Result<ObjectMeta>> = if self.is_collection() {
+            list_with_cache(ctx, store, &self.prefix).await?
+        } else {
+            match store.head(&self.prefix).await {
+                Ok(meta) => futures::stream::once(async { Ok(meta) })
+                    .map(|res| res.map_err(|e| DataFusionError::ObjectStore(Box::new(e))))
+                    .boxed(),
+                Err(_) => list_with_cache(ctx, store, &self.prefix).await?,
```

Comment:
Can we please add a comment here that explains the rationale for retry here. Something like 

```suggestion
                // If the head command fails, object doesn't exist. Retry as though
                // it were a prefix with LIST
                Err(_) => list_with_cache(ctx, store, &self.prefix).await?,
```

Also, I wonder if we should only retry on `Not found` errors, -- as it is this code retries on everything